### PR TITLE
feat: twitter shout-out and sponsor sections

### PR DESCRIPTION
**Pull request by n05la3**
Embedding tweets from twitter and styling them to conform to templates proved a little troublesome. 
- Firstly, I was unable to remove the label "Tweet your reply" on the embedded tweet. I didn't find anything official from the API on how to remove it.  
tweet-your-reply

- Secondly, the height of the embedded tweet depends on the width in which it is placed (this is set in the iframe of the embedded tweet). This means I cannot therefore control the height (to be as in the template design) of the the embedded tweet by setting a specific height (it will overflow) on its parent.

For these reasons, the resulting embeds do not really look exactly like the template design. Please take a look and give me your feedback. If there is a better approach, I'll create a fix.

### CODE REVIEW

**Comment by IlCallo** [comment]
> Firstly, I was unable to remove the label "Tweet your reply" on the embedded tweet. I didn't find anything official from the API on how to remove it.

It's ok, let's keep it like this, it's probably even better
Gonna check out other stuff asap

File: docs/src/assets/landing-page/image-links.ts [hunk 1]
```diff
@@ -1,10 +1,8 @@
-const sponsorLogos = [
-  'sponsor-logos/dreamonkey.png',
-  'sponsor-logos/comcomservices.png',
-  'sponsor-logos/project-finance.png',
-  'sponsor-logos/debricked.png',
-  'sponsor-logos/digitalocean.png'
-]
+const sponsorLogos = {
+  platinum: [ 'sponsor-logos/dreamonkey.png', 'sponsor-logos/comcomservices.png', 'sponsor-logos/project-finance.png' ],
```
Comment (IlCallo) [inline docs/src/assets/landing-page/image-links.ts]: You could extract the `sponsor-logos` part, since it's the same for all logos

File: docs/src/pages/landing-page/Index.vue [hunk 2]
```diff
@@ -134,13 +134,55 @@
           <div class="lp-heading lp-heading-large">
             What our community thinks of quasar
           </div>
-          <div class="text-size-20 text-lp-light text-weight-medium">
+          <div class="lp-heading lp-heading-small">
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: I'd use use BEM style `lp-heading--small` for the modifier class here and with similar classes

File: docs/src/pages/landing-page/Index.vue [hunk 3]
```diff
@@ -134,13 +134,55 @@
...
         </div>
       </div>
+      <q-carousel
+        v-model="slide"
+        transition-prev="scale"
+        transition-next="scale"
+        swipeable
+        animated
+        navigation
+        :autoplay="true"
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: Why is it needed to specify it as true? Isn't `autoplay` attribute ok?

Comment (IlCallo) [reply docs/src/pages/landing-page/Index.vue]: Also, if you make it `autoplay`, make it infinite too. But ask Nick if autoplay is actually needed

File: docs/src/pages/landing-page/Index.vue [hunk 4]
```diff
@@ -134,13 +134,55 @@
...
+        padding
+        arrows
+        height="400px"
+        class="bg-transparent rounded-borders"
+      >
+        <template v-slot:navigation-icon="{ active, btnProps, onClick }">
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: Use shorthand slot directive form: `#navigation-icon`

File: docs/src/pages/landing-page/Index.vue [hunk 5]
```diff
@@ -134,13 +134,55 @@
...
+          <q-btn
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: These navigation dots should actually be the same size AFAIK
Should also be easier to write the template

```vue
          <q-btn
            size="xs"
            :icon="btnProps.icon"
            :color="active ? 'lp-primary' : 'grey-8"
            flat
            round
            dense
            @click="onClick"
          />
```

File: docs/src/pages/landing-page/Index.vue [hunk 6]
```diff
@@ -134,13 +134,55 @@
...
+            color="grey-8"
+            flat
+            round
+            dense
+            @click="onClick"
+          />
+        </template>
+        <q-carousel-slide :name="i" class="showcase-cards" v-for="(showcaseContent, i) in showcaseContents" :key="i">
+          <div class="my-grid">
+            <div style="width: 90%" v-for="(showcases, i) in showcaseContent" :key="i">
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: Let aside the classic "don't use inline styles", why is this needed? You can control this via CSS grid properties if needed, but in this case it seems fine even if you just remove this inline style

Comment (IlCallo) [reply docs/src/pages/landing-page/Index.vue]: Try to use expanded and meaninful variable names where possible, in this case `index` instead of `i`
Also, you're shadowing the previous `i` iterator variable, and creating multiple element with the same key
Possibly differentiate them using template strings: ``:key="`slide-${slideIndex}`"`` and ``:key="`twitter-card-${cardIndex}`"``

Comment (IlCallo) [reply docs/src/pages/landing-page/Index.vue]: Why is this div needed, btw?
Same for the container div, you could just add `my-grid` into the `q-carousel-slide` tag

File: docs/src/pages/landing-page/Index.vue [hunk 7]
```diff
@@ -134,13 +134,55 @@
...
+              <twitter-card :card-content="showcases"/>
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: The `card-content` seems always the same except for the link, as it seems just a fallback
Can you explain how is this meant to work?

File: docs/src/pages/landing-page/Index.vue [hunk 8]
```diff
@@ -168,21 +210,87 @@
       <q-img class="absolute-bottom" width="50%" src="~assets/landing-page/homepage-background-images/astronaut-left-hand.png" />
       <q-img width="50%" class="absolute-top-right" src="~assets/landing-page/homepage-background-images/astronaut-right-hand.png" />
     </div>
+
+    <div class="text-center lg-mb-large">
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: I'm pretty sure `Full sponsor's list` button should bring the user to this section

File: docs/src/pages/landing-page/Index.vue [hunk 9]
```diff
@@ -168,21 +210,87 @@
...
+          <q-img :src="src" width="200px" v-for="(src, i) in sponsorLogos.platinum" :key="i" />
+          <div class="q-my-md">Gold Sponsors</div>
+          <q-img :src="src" width="200px" v-for="(src, i) in sponsorLogos.gold" :key="i" />
+          <div class="q-my-md">Silver Sponsors</div>
+          <q-img :src="src" width="200px" v-for="(src, i) in sponsorLogos.silver" :key="i" />
+        </div>
+      </div>
+    </div>
+
+    <div class="text-center final-words-section">
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: I'd change `final-words-section` with something like `social-channels-call-to-action` or similar, describe the semantic meaning, not what it is into the layout (unless you're writing a reusable component which act as a layout manager, which is not the case)

File: docs/src/pages/landing-page/Index.vue [hunk 10]
```diff
@@ -168,21 +210,87 @@
...
+      <q-img src="~assets\landing-page\homepage-background-images\planet.png">
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: I guess you want to change the backslashes here too

File: docs/src/pages/landing-page/Index.vue [hunk 11]
```diff
@@ -196,4 +304,22 @@ export default defineComponent({
 .support {
   height: 100vh;
 }
+
+.my-grid {
+  font-size: .5rem;
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: Any reason why you started using `rem`s here instead of pixels?

File: docs/src/pages/landing-page/Index.vue [hunk 12]
```diff
@@ -196,4 +304,22 @@ export default defineComponent({
...
+  display: grid;
+  grid-template-columns: repeat(3, 1fr);
+  grid-column-gap: 1rem;
+  grid-row-gap: 2rem;
+  align-content: center;
+}
+
+.final-words-section {
+  // TODO: reduce image height (from top), and remove this negative margin
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue:269]: Is this something Nick should take care?

File: docs/src/pages/landing-page/Index.vue [hunk 13]
```diff
@@ -196,4 +304,22 @@ export default defineComponent({
...
+  margin-top: -30%;
+}
+
+.showcase-cards {
+  overflow: hidden;
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: This doesn't seem to do much, what is it meant for?

File: docs/src/pages/landing-page/Index.vue [hunk 14]
```diff
@@ -168,21 +210,87 @@
...
+          '<blockquote class="twitter-tweet"><p lang="en" dir="ltr">We use <a href="https://twitter.com/quasarframework?ref_src=twsrc%5Etfw">@quasarframework</a> in some of our projects and we see that UI development using it is quite faster. </p>&mdash; MarsDevs (@marsdevs) <a href="https://twitter.com/marsdevs/status/1453670879825629189?ref_src=twsrc%5Etfw">October 28, 2021</a></blockquote>',
+          '<blockquote class="twitter-tweet"><p lang="en" dir="ltr">We use <a href="https://twitter.com/quasarframework?ref_src=twsrc%5Etfw">@quasarframework</a> in some of our projects and we see that UI development using it is quite faster. </p>&mdash; MarsDevs (@marsdevs) <a href="https://twitter.com/marsdevs/status/1453670879825629189?ref_src=twsrc%5Etfw">October 28, 2021</a></blockquote>'
+        ]
+      ],
       whyQuasar,
-      sponsorLogos
+      sponsorLogos,
+      slide: ref(0)
```
Comment (IlCallo) [inline docs/src/pages/landing-page/Index.vue]: Always prefer to define refs and other stuff needed by the template into the `setup` method, instead of inlining them into the returned object

File: docs/src/components/landing-page/TwitterCard.vue [hunk 15]
```diff
@@ -0,0 +1,39 @@
+<template>
+  <div v-html="cardContent">
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterCard.vue]: Unless incredibly rare scenarios, when you're using `v-html` you got something wrong, and this scenario doesn't seem to be an exception

File: docs/src/components/landing-page/TwitterCard.vue [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+  </div>
+</template>
+
+<script lang="ts">
+import { defineComponent } from 'vue'
+import { fabTwitter } from '@quasar/extras/fontawesome-v5'
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterCard.vue]: This doesn't seem to be used

File: docs/src/components/landing-page/TwitterCard.vue [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+export default defineComponent({
+  name: 'ShowCaseCard',
+  props: {
+    cardContent: {
+      required: true,
+      type: String
+    }
+  },
+  setup () {
+    const scriptElement = document.createElement('script')
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterCard.vue]: I'm pretty sure you only need one of these, not one per twitter card
Move this code into landing index component setup, or even better into the index html template, as an async/defer script

**Review by IlCallo — CHANGES_REQUESTED** [review]

File: docs/src/pages/landing-page/Index.vue (hunk 3 above)
Comment (n05la3) [reply docs/src/pages/landing-page/Index.vue]: the autoplay is totally an error from my part. I had some linting errors initially, I had put that there to see if it goes away, but then totally missed resetting it.

File: docs/src/pages/landing-page/Index.vue (hunk 12 above)
Comment (n05la3) [reply docs/src/pages/landing-page/Index.vue:269]: Yeah, I needed this PR done with so I can discuss the changes with Nick.

File: docs/src/pages/landing-page/Index.vue (hunk 13 above)
Comment (n05la3) [reply docs/src/pages/landing-page/Index.vue]: This is because of the responsiveness of the heights of the twitter cards, if for some reason a card (maybe the content of the tweet changed or so?) was larger than the height of the carousel specified then it will overflow and create horizontal scrollbars. Overflow hidden will prevent a weird the weird looking scrollbars.

File: docs/src/components/landing-page/TwitterCard.vue (hunk 15 above)
Comment (n05la3) [reply docs/src/components/landing-page/TwitterCard.vue]: I went this way so I could very easily place the content in an array an loop through them. I didn't really see this as bad since the data therein is not coming from a user and is static. I'll make the whole carousel a component (so it handles everything twitter carousel related), getting rid of v-html means I don't really need the twitter component anyways.

File: docs/src/pages/landing-page/Index.vue (hunk 7 above)
Comment (n05la3) [reply docs/src/pages/landing-page/Index.vue]: I have an array, whose elements are arrays of tweets (3 tweets per array). So I go through the larger array and display the 3 tweets per carousel slide.  So the content of card-content is not same through out. Once this is done, I'll talk with Nick to get all the real tweets that will go in there.

File: docs/src/pages/landing-page/Index.vue (hunk 8 above)
Comment (n05la3) [reply docs/src/pages/landing-page/Index.vue]: yeah, that would be correct. I'll in the next PR where I link to components page add that.

File: docs/src/pages/landing-page/Index.vue (hunk 7 above)
Comment (IlCallo) [reply docs/src/pages/landing-page/Index.vue]: Ok, but the common structure seems the same, is there a way to keep the structure as normal HTML and only pass-by the twitter id/link/whatever identifier you're using to load it later on?

File: docs/src/pages/landing-page/Index.vue (hunk 13 above)
Comment (IlCallo) [reply docs/src/pages/landing-page/Index.vue]: Strange, could not reproduce when I was testing it
Anyway, in these cases **always** add an explanation about why that's needed
Remember your code should always be understandable by a stranger (or yourself in 6 months, which is basically the same), add comments accordingly every time you add need to fix a strange interaction which isn't obvious

File: docs/src/components/landing-page/TwitterCard.vue (hunk 15 above)
Comment (IlCallo) [reply docs/src/components/landing-page/TwitterCard.vue]: > getting rid of v-html means I don't really need the twitter component anyways.
This is the main reason I said something is probably wrong

Having that code into a variable won't get you html highlight, wrong html tag closing detection, etc etc
Not bad per-se, but strips away a lot of safenets which the IDE provides you out of the box

Comment (n05la3) [reply docs/src/components/landing-page/TwitterCard.vue]: roger that.

File: docs/src/components/landing-page/TwitterCard.vue (hunk 17 above)
Comment (n05la3) [reply docs/src/components/landing-page/TwitterCard.vue]: Strangely, when I move it to the index template (as async or defered) or to the Index.vue or to the new component (TwitterShowcaseCards) which loads the TwitterCard compnent, only the first set of tweets load, the rest of the tweets in subsequent slides do not load. I'm yet to figure out what the issue is, I have however updated this PR (with fixes to issues you pointed out), and with this issue remaining issue commented, to give you a view.

I was also looking at  displaying a loading state when a tweet is being loaded. Coupled with the first issue, I may have to go with rendering with Js, which renders the tweets as a promise.

File: docs/src/pages/landing-page/Index.vue (hunk 11 above)
Comment (n05la3) [reply docs/src/pages/landing-page/Index.vue]: So I'm thinking in situations where body's 1rem is 16px that should equal about 8px. And in situations where it is say less, that should scale down automatically. I will move it back to the px to adhere to the px style so far. And handle such responsiveness later on.

File: docs/src/components/landing-page/TwitterCard.vue (hunk 17 above)
Comment (IlCallo) [reply docs/src/components/landing-page/TwitterCard.vue]: Ok, let me know if you need help to understand what's going on there. Copy here the relevant resources you're using to add this feature, in you need help

Comment (n05la3) [reply docs/src/components/landing-page/TwitterCard.vue]: In order to embed the cards, one needs the script at `https://platform.twitter.com/widgets.js ` to be loaded in the head section of the page. Then add the code below to embed a tweet with id tweetId somewhere in the page. 
```
<blockquote class="twitter-tweet">
  <a :href="`https://twitter.com/marsdevs/status/${tweetId}`"></a>
</blockquote>
```
In order to inject the script at `https://platform.twitter.com/widgets.js` into the page, I have used the script below to do so 
```
scriptElement.setAttribute('src', 'https://platform.twitter.com/widgets.js')
scriptElement.setAttribute('charset', 'utf-8')
document.head.appendChild(scriptElement)
```
I added two extra components, one which displays the carousel (TwitterShowcaseCards.vue) and another which displays a twitter card (TwitterCard.vue).

I inject the head script from within  TwitterShowcaseCards.vue (this script will be run only once now, not for every card). But unfortunately only the first slide renders. If I inject the script from within TwitterCard.vue (where the script is called for every card), then the rest of the slides render correctly.

So at the moment, I have also tried injecting directly in the head tag of index.template.html, and the behaviour is exactly the same as injecting in TwitterShowcaseCards.vue (where only first slide renders).

Comment (IlCallo) [reply docs/src/components/landing-page/TwitterCard.vue]: This is due to the twitter widget searching the tweets to embed only the first time you include it
You should use dynamic initialization upon slide change

image

File: docs/src/components/landing-page/TwitterShowcaseCards.vue [hunk 18]
```diff
@@ -36,65 +37,82 @@
...
         </div>
       </div>
     </q-carousel-slide>
   </q-carousel>
 </template>
 
-<script>
-import { defineComponent, ref } from 'vue'
-import TwitterCard from 'src/components/landing-page/TwitterCard'
+<script lang="ts">
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterShowcaseCards.vue]: We aren't using TS in Quasar codebase

File: docs/src/components/landing-page/TwitterShowcaseCards.vue [hunk 19]
```diff
@@ -36,65 +37,82 @@
...
+  })
+}
+
 export default defineComponent({
   name: 'TwitterShowcaseCards',
-  components: { TwitterCard },
   setup () {
     const slide = ref(0)
+    let twttr = null
+    const groupIsDisplayed = []
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterShowcaseCards.vue]: A better name for this would be `alreadyDisplayedTweetGroupsIndexes`
Sure it's long, but its descriptive. When you need to decide between length and descriptiveness, prefer the latter

File: docs/src/components/landing-page/TwitterShowcaseCards.vue [hunk 20]
```diff
@@ -36,65 +37,82 @@
...
+<script lang="ts">
+import { defineComponent, onMounted, ref } from 'vue'
 
-// TODO: fix the issue causing ONLY the first tweet group (slide 1) display
 const scriptElement = document.createElement('script')
 scriptElement.setAttribute('src', 'https://platform.twitter.com/widgets.js')
 scriptElement.setAttribute('charset', 'utf-8')
 document.head.appendChild(scriptElement)
 
+const addTwttr = async () => {
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterShowcaseCards.vue]: I'd use `getTwitterInstance` or similar

File: docs/src/components/landing-page/TwitterShowcaseCards.vue [hunk 21]
```diff
@@ -36,65 +37,82 @@
...
+          conversation: 'none',
+          cards: 'hidden'
+        }
+      )
+    }
+
+    onMounted(async () => {
+      twttr = await addTwttr()
+      // display first tweet group
+      tweetGroups[ 0 ].forEach(tweetId => {
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterShowcaseCards.vue]: It doesn't change anything feature-wise, but it's much better to create a constant outside the component scope as `INITIAL_SLIDE_INDEX` or `INITIAL_TWEET_GROUP_INDEX` and use it both to initialize `slide` ref, to access `tweetGroups` and to push it into the array of groups indexes already displayed
It improves readability and makes sure anyone coming by afterwards knows those indexes are related and must all be changed together

File: docs/src/components/landing-page/TwitterShowcaseCards.vue [hunk 22]
```diff
@@ -36,65 +37,82 @@
...
+        createTweetCard(twttr, tweetId, `tweet-container-${tweetId}`)
+      })
+      groupIsDisplayed.push(0)
+    })
+
+    const loadTweets = () => {
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterShowcaseCards.vue]: Prefer using `function` over declaring a `const`, unless the given function is really simple and directly returns a value (eg. wrapper fuctions, mapping functions, etc)
Here you don't spare a single char, so just use `function` which is what most people would expect to be used

File: docs/src/components/landing-page/TwitterShowcaseCards.vue [hunk 23]
```diff
@@ -36,65 +37,82 @@
...
       ],
       [
-        '1453670879825629189',
-        '1453670879825629189',
-        '1453670879825629189'
+        '1419939610067623937'
       ]
     ]
 
+    const createTweetCard = (twttr, tweetId, tweetContainerId) => {
```
Comment (IlCallo) [inline docs/src/components/landing-page/TwitterShowcaseCards.vue]: Prefer using `function` over declaring a `const`, unless the given function is really simple and directly returns a value (eg. wrapper fuctions, mapping functions, etc)
Here you don't spare a single char, so just use `function` which is what most people would expect to be used

**Review by IlCallo — CHANGES_REQUESTED** [review]

File: docs/src/components/landing-page/TwitterShowcaseCards.vue (hunk 20 above)
Comment (IlCallo) [reply docs/src/components/landing-page/TwitterShowcaseCards.vue]: You should use `function` here too, or remove the code block and just return the promise
Using `function` is preferred tho
